src/main/java/com/iota/iri/network/pipeline/ReplyStage.java

```diff
@@ -0,0 +1,167 @@
+package com.iota.iri.network.pipeline;
+
+import com.iota.iri.conf.NodeConfig;
+import com.iota.iri.controllers.TipsViewModel;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.model.HashFactory;
+import com.iota.iri.network.FIFOCache;
+import com.iota.iri.network.NeighborRouter;
+import com.iota.iri.network.neighbor.Neighbor;
+import com.iota.iri.network.protocol.Protocol;
+import com.iota.iri.service.milestone.LatestMilestoneTracker;
+import com.iota.iri.service.snapshot.SnapshotProvider;
+import com.iota.iri.storage.Tangle;
+
+import java.security.SecureRandom;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ReplyStage} replies to the neighbor which supplied the given hash of the requested transaction. If a
+ * {@link Hash#NULL_HASH} is supplied, then a random tip is replied back to the neighbor. A neighbor indicates to
+ * receive a random tip, when the requested transaction hash is the same as the transaction hash of the transaction in
+ * the gossip payload.
+ */
+public class ReplyStage implements Stage {
+
+    private static final Logger log = LoggerFactory.getLogger(ReplyStage.class);
+
+    private NeighborRouter neighborRouter;
+    private Tangle tangle;
+    private NodeConfig config;
+    private TipsViewModel tipsViewModel;
+    private LatestMilestoneTracker latestMilestoneTracker;
+    private SnapshotProvider snapshotProvider;
+    private FIFOCache<Long, Hash> recentlySeenBytesCache;
+    private SecureRandom rnd = new SecureRandom();
```

Comment:
Why do you have a `Random` in a test?
I think tests should be deterministic as much as possible.

There are some tests in IRI that use `Random` but that is not good.
In face I just opened an issue about that:
https://github.com/iotaledger/iri/issues/1497

Reply:
`ReplyStage.java` is not a test class and if you check where `rnd` is used it is in the bottom part of the file for either loading a random tip or the latest milestone hash:
```
    private Hash getRandomTipPointer() {
        Hash tip = rnd.nextDouble() < config.getpSendMilestone() ? latestMilestoneTracker.getLatestMilestoneHash()
                : tipsViewModel.getRandomSolidTipHash();
        return tip == null ? Hash.NULL_HASH : tip;
    }
```

In the actual test file `ReplyStageTest.java`, there's a **mocked** `SecureRandom` instance where I do `Mockito.when(random.nextDouble()).thenReturn(0.6d);` to make the test deterministic.

Did you encountered errors while running the tests I wrote maybe?